File: py/candlestick_patterns/patterns/homing_pigeon.py
"""Homing Pigeon pattern (2-candle bullish)."""
from __future__ import annotations

from ..core.primitives import is_black, real_body
from ...fuzzy import t_product_all


def homing_pigeon(self) -> float:
    """Homing Pigeon: a two-candle bullish pattern.

    Must have:
    - first candle: long black,
    - second candle: short black, real body engulfed by first candle's
      real body.

    The meaning of "long" is specified with self._long_body.
    The meaning of "short" is specified with self._short_body.

    Category A: always bullish (continuous).

    Returns:
        Continuous float in [0, 100].  Always bullish.
    """
    if not self._enough(2, self._long_body, self._short_body):
        return 0.0

    o1, h1, l1, c1 = self._bar(2)
    o2, h2, l2, c2 = self._bar(1)

    # Crisp gates: both black.
    if not (is_black(o1, c1) and is_black(o2, c2)):
        return 0.0

    # Fuzzy conditions.
    mu_long1 = self._mu_greater(real_body(o1, c1), self._long_body, 2)
    mu_short2 = self._mu_less(real_body(o2, c2), self._short_body, 1)

    # Containment: second body engulfed by first body.
    # For black candles: open > close, so upper = open, lower = close.
    eq_width = self._fuzz_ratio * self._avg(self._equal, 2)
    mu_enc_upper = self._mu_lt_raw(o2, o1, eq_width)
    mu_enc_lower = self._mu_gt_raw(c2, c1, eq_width)

    confidence = t_product_all(mu_long1, mu_short2, mu_enc_upper, mu_enc_lower)

    return confidence * 100.0

File: ts/candlestick-patterns/patterns/homing-pigeon.ts
/** Homing Pigeon pattern (2-candle bullish). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, isBlack } from '../core/primitives.ts';

/**
 * Homing Pigeon: a two-candle bullish pattern.
 *
 * Must have:
 * - first candle: long black,
 * - second candle: short black, real body engulfed by first candle's
 *   real body.
 *
 * The meaning of "long" is specified with longBody.
 * The meaning of "short" is specified with shortBody.
 *
 * Category A: always bullish (continuous).
 *
 * Returns:
 *     Continuous number in [0, 100]. Always bullish.
 */
export function homingPigeon(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(2, cp.longBody, cp.shortBody)) return 0.0;

    const b1 = cp.bar(2);
    const b2 = cp.bar(1);

    // Crisp gates: both black.
    if (!(isBlack(b1.o, b1.c) && isBlack(b2.o, b2.c))) return 0.0;

    // Fuzzy conditions.
    const muLong1 = cp.muGreaterCS(realBody(b1.o, b1.c), cp.longBody, 2);
    const muShort2 = cp.muLessCS(realBody(b2.o, b2.c), cp.shortBody, 1);

    // Containment: second body engulfed by first body.
    // For black candles: open > close, so upper = open, lower = close.
    const eqWidth = cp.fuzzRatio * cp.avgCS(cp.equal, 2);
    const muEncUpper = cp.muLtRaw(b2.o, b1.o, eqWidth);
    const muEncLower = cp.muGtRaw(b2.c, b1.c, eqWidth);

    return tProductAll(muLong1, muShort2, muEncUpper, muEncLower) * 100.0;
}
